contrib/sarplus/README.md

```diff
@@ -0,0 +1,191 @@
+# sarplus (preview)
+pronounced sUrplus as it's simply better if not best!
+
+[![Build Status](https://dev.azure.com/best-practices/recommenders/_apis/build/status/contrib%20sarplus?branchName=master)](https://dev.azure.com/best-practices/recommenders/_build/latest?definitionId=107&branchName=master)
+[![PyPI version](https://badge.fury.io/py/pysarplus.svg)](https://badge.fury.io/py/pysarplus)
+
+Features
+* Scalable PySpark based [implementation](python/pysarplus/SARPlus.py)
+* Fast C++ based [predictions](python/src/pysarplus.cpp)
+* Reduced memory consumption: similarity matrix cached in-memory once per worker, shared accross python executors 
+* Easy setup using [Spark Packages](https://spark-packages.org/package/Microsoft/sarplus)
+
+# Benchmarks
+
+| # Users | # Items | # Ratings | Runtime | Environment | Dataset | 
+|---------|---------|-----------|---------|-------------|---------|
+| 2.5mio  | 35k     | 100mio    | 1.3h    | Databricks, 8 workers, [Azure Standard DS3 v2](https://azure.microsoft.com/en-us/pricing/details/virtual-machines/linux/) (4 core machines) | |
+
+# Top-K Recommendation Optimization
+
+There are a couple of key optimizations:
+
+* map item ids (e.g. strings) to a continuous set of indexes to optmize storage and simplify access
+* convert similarity matrix to exactly the representation the C++ component needs, thus enabling simple shared, memory mapping of the cache file and avoid parsing. This requires a customer formatter, written in Scala
+* shared read-only memory mapping allows us to re-use the same memory from multiple python executors on the same worker node
+* partition the input test users and past seen items by users, allowing for scale out
+* perform as much of the work as possible in PySpark (way simpler)
+* top-k computation
+** reverse the join by summing reverse joining the users past seen items with any related items
+** make sure to always just keep top-k items in-memory
+** use standard join using binary search between users past seen items and the related items
+
+![Image of sarplus top-k recommendation optimization](images/sarplus_udf.svg) 
+
+# Usage
+
+```python
+import pandas as pd
+from pysarplus import SARPlus
+
+# spark dataframe with user/item/rating/optional timestamp tuples
+train_df = spark.createDataFrame(
+      pd.DataFrame({

```

Comment:
is pandas only used to generate the dataframe?
if so you could remove the import and create the dataframes by using a list of tuples
```
train_df = spark.createDataFrame([(1, 1, 1), (1, 2, 1), (2, 1, 1), (3, 1, 1), (3, 3, 1)], ['user_id', 'item_id', 'rating'])
```